src/mongo/client/gridfs.cpp

```diff
@@ -321,7 +321,13 @@ namespace mongo {
             invariant( _pendingDataSize <= _chunkSize );
             if (_pendingDataSize == _chunkSize) {
                 _appendPendingData();
-                _appendChunk( data + size, length - size, false );
+                const char* const end = data + length;
+                data = _appendChunk( data + size, length - size,false);
+                if (data != end) {
+                    size_t nsize = static_cast<size_t>(end-data);
```

Comment:
spaces around arithmetic op. Also, consider invariant(end > data), maybe before entering the if block.